Only wrap descendants with render method, closes #84

nfcampos opened the pull request:
closes #84 (see the issue for a discussion of the changes)

## todo
- [x] fix current tests
- [x] make change to index.js
- [x] write additional tests

nfcampos commented:
@gaeron should i write a new test for each of the cases the current tests used to cover?

gaearon commented:
Please do.

nfcampos commented:
something like this is not supposed to be wrapped, is it?
```js
import React, { Component } from 'react'
class Foo extends Component {
  render() {}
}
```

gaearon commented:
This looks like a React component to me, why not?

nfcampos commented:
because it requires changing this line https://github.com/gaearon/babel-plugin-react-transform/blob/master/src/index.js#L174
to `superClasses: options.superClasses || ['React.Component', 'Component'],
which mean that any class that extends a class called `Component` and has a render method is wrapped.
is that ok? or is there some other way of doing this?

gaearon commented:
Good catch. I would say let’s add `Component` to defaults. It’s a too common pattern.

nfcampos commented:
yeah i agree. does that go into this PR or another one (it's not really related to this)?

gaearon commented:
It’s fine to do this here.

nfcampos commented:
Alright, it's all done i think.
changes:
- only wrap classes that extend the correct superClass *and* have a render method
- changed tests to reflect change above
- added `Component` to default superClasses
- added test for the above

gaearon commented:
Seems like technically this wouldn’t even be a breaking change because we used to allow any class with `render()` anyway.

nfcampos commented:
well it breaks some stuff that shouldn't work anyways, so that part is fine

to me it looks like the only breaking change would be stuff like this:
```js
import React from 'react'
const Comp = React.Component
class Foo extends Comp {
  render() {}
}
```
this would have worked previously because it had a render method and now won't, but would anyone actually do something like this?

gaearon commented:
Even if they used this pattern, we’re not technically “breaking” it because the code would still work. Transforms are meant for development only anyway.

nfcampos commented:
yeah that's true

jwbay commented on src/index.js:
```diff
@@ -119,7 +119,7 @@ export default function({ types: t, template }) {
       if (
         path.node[VISITED_KEY] ||
-        !matchesPatterns(path.get('callee'), this.factoryMethods) &&
+        !matchesPatterns(path.get('callee'), this.factoryMethods) ||
```
if both `matchesPatterns` and `isReactLikeComponentObject` are independent checks for "is this node a react component", then we should only bail and return early if *both* the checks fail, right? if a component would pass the `isReactLikeComponentObject` check but fails the `factoryMethods` check, it never gets a chance to do so:
`if (false || true || false)` -> bail

gaearon commented:
Out in 2.0.1, thank you a lot!

nfcampos commented:
Happy to help!

gaearon commented:
If you have time, I’d appreciate you looking into why https://github.com/gaearon/babel-plugin-react-transform/pull/87 fails, merging/updating it if it’s correct, and checking whether it fixes another correctness issue: https://github.com/gaearon/babel-plugin-react-transform/issues/88.

nfcampos commented:
I'll look into it
